File: babydata/src/main/java/com/zero/babydata/room/BabyDatabase.java
package com.zero.babydata.room;

import android.content.Context;

import androidx.room.Database;
import androidx.room.Room;
import androidx.room.RoomDatabase;

import com.zero.babydata.entity.BabyInfo;
import com.zero.babydata.entity.ChildDailyRecord;
import com.zero.babydata.entity.EventRecord;
import com.zero.babydata.entity.FeedingRecord;
import com.zero.babydata.entity.SleepRecord;

import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

@Database(entities = {BabyInfo.class, FeedingRecord.class, SleepRecord.class, ChildDailyRecord.class, EventRecord.class}, version = 1)
public abstract class BabyDatabase extends RoomDatabase {

    private static BabyDatabase instance;

    public abstract BabyInfoDao babyInfoDao();

    public abstract FeedingRecordDao feedingRecordDao();

    public abstract SleepRecordDao sleepRecordDao();

    public abstract ChildDailyRecordDao childDailyRecordDao();

    public abstract EventRecordDao eventRecordDao();

    public static synchronized BabyDatabase getInstance(Context context) {
        if (instance == null) {
            instance = Room.databaseBuilder(context.getApplicationContext(),
                            BabyDatabase.class, "baby_database")
                    .fallbackToDestructiveMigration()  // Handle migration automatically
                    .build();
        }
        return instance;
    }

    private static final ExecutorService databaseWriteExecutor = Executors.newFixedThreadPool(4);

    public static ExecutorService getDatabaseWriteExecutor() {
        return databaseWriteExecutor;
    }
}

File: babydata/src/main/java/com/zero/babydata/room/BabyDatabase.kt
package com.zero.babydata.room

import android.content.Context
import androidx.room.Database
import androidx.room.Room
import androidx.room.RoomDatabase
import com.zero.babydata.entity.BabyInfo
import com.zero.babydata.entity.ChildDailyRecord
import com.zero.babydata.entity.EventRecord
import com.zero.babydata.entity.FeedingRecord
import com.zero.babydata.entity.SleepRecord
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors

@Database(
    entities = [
        BabyInfo::class,
        FeedingRecord::class,
        SleepRecord::class,
        ChildDailyRecord::class,
        EventRecord::class
    ],
    version = 4,  // v4: EventRecord 添加 endTime, extraData 字段
    exportSchema = true  // 商业应用建议开启，便于追踪数据库变更
)
abstract class BabyDatabase : RoomDatabase() {

    abstract fun babyInfoDao(): BabyInfoDao
    abstract fun feedingRecordDao(): FeedingRecordDao
    abstract fun sleepRecordDao(): SleepRecordDao
    abstract fun childDailyRecordDao(): ChildDailyRecordDao
    abstract fun eventRecordDao(): EventRecordDao

    companion object {
        @Volatile
        private var instance: BabyDatabase? = null

        private val databaseWriteExecutor: ExecutorService = Executors.newFixedThreadPool(4)

        @JvmStatic
        fun getInstance(context: Context): BabyDatabase {
            return instance ?: synchronized(this) {
                instance ?: Room.databaseBuilder(
                    context.applicationContext,
                    BabyDatabase::class.java,
                    "baby_database"
                )
                    .allowMainThreadQueries()
                    // 商业应用禁止使用 fallbackToDestructiveMigration()
                    // 必须通过 Migration 保护用户数据
                    .addMigrations(*DatabaseMigrations.getAllMigrations())
                    .build()
                    .also { instance = it }
            }
        }

        @JvmStatic
        fun getDatabaseWriteExecutor(): ExecutorService = databaseWriteExecutor
    }
}
